BOOKKEEPER-588 SSL Support

eolivelli opened the pull request.

jvrao commented:
Kishore is reviewing it. I hope he can make this mornings call. He is on vacation and supposed to be back today.

kishorekasi commented on bookkeeper-server/nb-configuration.xml:
```diff
@@ -1,18 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
remove this netbeans relevant config file?

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/AuthProviderFactoryFactory.java:
```diff
@@ -25,10 +25,6 @@
 import org.apache.bookkeeper.client.BKException;
 import org.apache.bookkeeper.conf.ClientConfiguration;
 import org.apache.bookkeeper.conf.ServerConfiguration;
-<<<<<<< HEAD
```
Fix merge issues

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java:
```diff
@@ -59,11 +59,7 @@
          *                   is complete.
          */
         BookieAuthProvider newProvider(BookieConnectionPeer connection,
-<<<<<<< HEAD
```
Same here. Fix the merge issues in this file too

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java:
```diff
@@ -284,4 +289,46 @@ public AbstractConfiguration setClientAuthProviderFactoryClass(
+    /**
+     * Set the list of enabled SSL protocols. leave blank not to override default JDK list
+     *
+     * @param list comma separated list of enabled SSL cipher suites
```
fix comment?

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java:
```diff
@@ -284,4 +289,46 @@ public AbstractConfiguration setClientAuthProviderFactoryClass(
+    /**
+     * Get the list of enabled SSL cipher suites
+     *
+     * @return the client authentication provider factory class name or null.
```
fix comment

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java:
```diff
@@ -284,4 +289,46 @@ public AbstractConfiguration setClientAuthProviderFactoryClass(
+    /**
+     * Get the list of enabled SSL protocols
+     *
+     * @return the client authentication provider factory class name or null.
```
fix comment

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java:
```diff
@@ -101,6 +101,31 @@
+    protected final static String CLIENT_SSL_AUTHENTICATION = "clientSSLAuthentication";
+
+    /**
+     * Key store which holds the client-side SSL certificate. The certificate must be in PKCS12 format
```
Any reason why are assuming only pkcs#12 format. Why not JKS or pkcs#8?

eolivelli replied:
This comes from the original patch from @ivankelly. In production I usually use PKCS12. I can make a new option for the format. Do you think it wild be useful?

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java:
```diff
@@ -882,6 +907,112 @@ public ClientConfiguration setBookieErrorThresholdPerInterval(long thresholdPerI
+
+    /**
+    * Specify whether the client validates SSL certificates
+    * @return whether the client will verify certificates or not.
```
fix comment

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java:
```diff
@@ -882,6 +907,112 @@ public ClientConfiguration setBookieErrorThresholdPerInterval(long thresholdPerI
+    /**
+      * Specify whether the client will send an SSL certificate on TLS-handshake
+      * @param Whether to send a certificate or not
+      * @return the server configuration
```
fix comment "server" -> "client"

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java:
```diff
@@ -882,6 +907,112 @@ public ClientConfiguration setBookieErrorThresholdPerInterval(long thresholdPerI
+      * @see #setClientSSLAuthentication(boolean)
+      * @return whether SSL is enabled on the bookie or not.
+      */
+     public boolean isClientSSLAuthentication() {
```
Reword function name or keep it consistent with the above convention of "set" or "get"

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java:
```diff
@@ -101,6 +101,31 @@
+    /**
+     * Key store which holds the client-side SSL certificate. The certificate must be in PKCS12 format
+     */
+    protected final static String CLIENT_SSL_KEY_STORE = "clientSSLKeyStore";
```
For local CA signed certificates, should we also have an option to provide the trustStore along with client/bookie keyStore?

eolivelli replied:
IMHO I think that check be done in the AuthProvider.

But we can take care of this directly here.
What do you think ?

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java:
```diff
@@ -70,6 +70,11 @@
 
+    // Common SSL configuration
+    protected final static String SSL_ENABLED_CIPHER_SUITES = "sslEnabledCipherSuites";
+    protected final static String SSL_ENABLED_PROTOCOLS = "sslEnabledProtocols";
```
Please provide more information in the comment about what protocols means here. Do you mean protocol version here?

eolivelli replied:
the option will be reflected to SSLEngine#setEnableProtocols.
I will update the comments on setters and here

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java:
```diff
@@ -0,0 +1,187 @@
+        return engine;
+    }
+
+    private TrustManager[] getTrustManagers() {
```
To keep more secure and enable both client/bookie trust only certificates from trusted issuers, should we use TrustStore configuration and create a trust manager from the provided truststore?

eolivelli replied:
IMHO I think that check be done in the AuthProvider.

But we can take care of this directly here.
What do you think ?

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java:
```diff
@@ -0,0 +1,187 @@
+    public SSLEngine getEngine() {
+        SSLEngine engine = ctx.createSSLEngine();
+        engine.setUseClientMode(isClient);
+        if (!isClient) {
```
Allowing both client and server to authenticate itself will enabled MutualTLS

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java:
```diff
@@ -79,16 +75,19 @@ default void close() {}
     }
 
     /**
+     * Callback to let the provider know that the underlying protocol is changed.
```
Is there a specific action that is needed as part of this callback in AuthProvider?

eolivelli replied:
I think this is a part of a previous commit, that has already been merged on master.
The idea is that the AuthProvider will have access to the Certificates, in order to provide custom banning/revoking and automatic close of connection after certificate expiration

eolivelli replied:
the AuthProvider on this callback has access to the Certificates.
The AuthProvider is instantiated when the connection is created, but no StartTLS as occurred yet, so at that time it is not possible to access client-side Certificates.
So this callback is useful to intercept the client SSL data and:
- schedule a close of the connection at the time of certificate expiration date
- prevent "forbidden" certificates to access Bookie data

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ConnectionPeer.java:
```diff
@@ -60,4 +60,10 @@
     public void setAuthorizedId(BookKeeperPrincipal principal);
+
+    /**
+     * This flag returns two if a 'secure' channel in use, like SSL
```
fix comment. "two" -> "true"

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java:
```diff
@@ -1712,4 +1716,74 @@ public void setBookieAuthProviderFactoryClass(String factoryClass) {
+      * @see #setSSLKeyStore(String)
+      * @return the server configuration
+      */
+     public ServerConfiguration setSSLKeyStorePassword(String keyStorePassword) {
```
Will it be useful to have the keyStore encryption password to be store in a file and this config parameter to accept the path?

eolivelli replied:
I have no strong opinion on this. Usually other common libraries use properties to store directly passwords and no path to files.
If you need a path to file I can change it.

b.q.
will the access to an absolute file system path  lead to other security concerns ?

kishorekasi commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java:
```diff
@@ -34,7 +34,6 @@
 import org.apache.bookkeeper.conf.ClientConfiguration;
```
As identified test cases are missing. Adding few more test cases to the list -

1. Multiple bookkeeper clients (a few with ssl enabled with multiple bookies (all ssl enabled)
2. Multiple bookkeeper clients (a few with ssl enabled with multiple bookies (a few with ssl disabled)

eolivelli replied:
I have added:
- TestSSL#testConnectToSSLClusterMixedClient
- TestSSL#testClientWantsButOnlySSLSomeServerHaveIt

kishorekasi commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java:
```diff
@@ -344,7 +335,8 @@ void connectIfNeededAndDoOp(GenericCallback<PerChannelBookieClient> op) {
-                    if (state == ConnectionState.CONNECTING) {
+                    if (state == ConnectionState.CONNECTING
+                        || state == ConnectionState.STARTTLS) {
                         // just return as connection request has already send
```
reword comment

eolivelli commented:
@ kishorekasi  thank you for your review.
I have addressed most of the issues and rebased to current master.

I have not run any benchmark on SSL, so I have no number to share about performance degradation

sijie commented:
@kishorekasi do you mind reviewing the new patch again?

sijie commented:
@eolivelli since netty 4 is in, do you mind rebasing this pull request?

eolivelli commented:
@sijie I can rebase, but @kishorekasi told he would have picked this issue up in order to implement more security features.
I am ok for rebasing, getting this merged and then let Kishore or @jvrao create another issue to complete the feature.
A new issue for docs needs to be created as well

sijie commented:
@kishorekasi @jvrao ^^ any thoughts?

eolivelli commented:
@sijie @jvrao @kishorekasi 
I will wait for your ACK before starting the rebase I think I will not be trivial

kishorekasi commented:
Enrico,

I have generate a pull request combing both our changes.

Kishore

Sent from my iPhone

> On May 17, 2017, at 8:29 AM, Enrico Olivelli <notifications@github.com> wrote:
> 
> @sijie @jvrao @kishorekasi
> I will wait for your ACK before starting the rebase I think I will not be trivial
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

kishorekasi commented:
Sorry, typo. I will generate a pull request.

eolivelli commented:
@kishorekasi this sounds great to me, I will close this PR once you open yours

sijie commented:
@kishorekasi any thoughts about the new SSL pull requests?

kishorekasi commented:
Hi Sijie,

I have some more testing to complete. I will work on it this weekend and post a pull request.

Kishore



Sent from my iPhone

> On May 24, 2017, at 5:07 PM, Sijie Guo <notifications@github.com> wrote:
> 
> @kishorekasi any thoughts about the new SSL pull requests?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

eolivelli commented:
closing this PR in favour of #183 from @kishorekasi
